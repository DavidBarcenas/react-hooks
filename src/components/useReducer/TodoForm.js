import React from 'react'
import { useForm } from '../../hooks/useForm'

export const TodoForm = ({handleAddTodo}) => {

  const [{desc}, handleInputchange, reset] = useForm({ desc: '' })

  const handleSubmit = (e) => {
    e.preventDefault();

    if(desc.trim().length <= 1) { return; }

    const newTodo = { id: new Date().getTime(), desc, done: false }
    
    handleAddTodo(newTodo);
    reset();
  }

  return (
    <form className="mb-3" onSubmit={handleSubmit}>
      <input type="text" name="desc" placeholder="ToDo" value={desc} onChange={handleInputchange}/>
      <button type="submit" className="btn btn-success btn-sm mb-1">Add ToDo</button>
    </form>
  )
}
