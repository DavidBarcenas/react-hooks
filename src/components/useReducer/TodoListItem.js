import React from 'react'
import PropTypes from 'prop-types'

export const TodoListItem = ({todo, i, handleToogle, handleDelete }) => {
  return (
    <li className="list-group-item d-flex justify-content-between align-items-center">
      <label className={`${todo.done && 'complete'} mb-0`} onClick={() => handleToogle(todo.id)}>{i + 1} - {todo.desc}</label>
      <button className="btn btn-danger btn-sm" onClick={() => handleDelete(todo.id)}>X</button>
    </li>
  )
}

TodoListItem.propTypes = {
  todo: PropTypes.object.isRequired,
  i: PropTypes.number.isRequired,
  handleToogle: PropTypes.func.isRequired,
  handleDelete: PropTypes.func.isRequired,
}