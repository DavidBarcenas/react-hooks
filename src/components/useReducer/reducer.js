const initialState = [{
  id: 1,
  todo: 'Create app',
  done: false
}]


const todoReducer = (state = initialState, action) => {

  if(action.type=== 'agregar') {
    return [...state, action.payload]
  }

  return state;
}

let todos;

const newTodo = {
  id: 2,
  todo: 'Add wallet',
  done: false
}

const action = {
  type: 'agregar',
  payload: newTodo
}

todos = todoReducer(todos, action)


console.log(todos)