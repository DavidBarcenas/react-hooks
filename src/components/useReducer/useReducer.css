.list-group-item {
  cursor: pointer;
}

.complete {
  text-decoration: line-through;
}