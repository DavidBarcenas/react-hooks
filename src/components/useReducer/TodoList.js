import React from 'react'
import PropTypes from 'prop-types';
import { TodoListItem } from './TodoListItem'

export const TodoList = ({todos, handleToogle, handleDelete}) => {
  return (
    <ul className="list-group text-left">
      {
        todos.map((todo, i) => 
          <TodoListItem 
            key={todo.id} 
            todo={todo} 
            i={i} 
            handleToogle={handleToogle} 
            handleDelete={handleDelete} />
          )
      }
    </ul>
  )
}

TodoList.propTypes = {
  todos: PropTypes.array.isRequired,
  handleToogle: PropTypes.func.isRequired,
  handleDelete: PropTypes.func.isRequired,
}