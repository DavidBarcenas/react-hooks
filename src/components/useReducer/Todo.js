import React, { useReducer, useEffect } from 'react'
import { todoReducer } from './todoReducer'
import { TodoList } from './TodoList'
import { TodoForm } from './TodoForm'
import './useReducer.css'

// const initialState = [{ id: new Date().getTime(), desc: 'Learning useReducer', done: false }]
const init = () => JSON.parse(localStorage.getItem('todos')) || [];

export const Todo = () => {

  const [todos, dispatch] = useReducer(todoReducer, [], init)

  useEffect(() => {
    localStorage.setItem('todos', JSON.stringify(todos))
  }, [todos])

  const handleDelete  = (todoID) => dispatch({ type: 'delete', payload: todoID });
  const handleToogle  = (todoID) => dispatch({ type: 'toggle', payload: todoID });
  const handleAddTodo = (newTodo) => dispatch({ type: 'add', payload: newTodo });

  return (
    <div className="text-center p-5">
      <h1 className="text-primary">Todo App ({todos.length})</h1>
      <hr/>

      <TodoForm handleAddTodo={handleAddTodo} />

      <TodoList 
        todos={todos}
        handleToogle={handleToogle}
        handleDelete={handleDelete} 
      />
    </div>
  )
}
