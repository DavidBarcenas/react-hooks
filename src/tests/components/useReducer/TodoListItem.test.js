import '@testing-library/jest-dom';
import React from 'react';
import { shallow } from 'enzyme';
import { demoTodos } from '../../fixtures/demoTodos';
import { TodoListItem } from '../../../components/useReducer/TodoListItem';

describe('<TodoListItem /> testing', () => {

  const handleToogle = jest.fn()
  const handleDelete = jest.fn()
  
  const wrapper = shallow(
    <TodoListItem 
      todo={demoTodos[0]} 
      i={demoTodos[0].id} 
      handleToogle={handleToogle} 
      handleDelete={handleDelete} 
    />)
  
  
  test('should render correctly', () => {
    const index = wrapper.find('label').text().split('-')[0]
    expect(wrapper).toMatchSnapshot()
    expect(index - 1).toBe(demoTodos[0].id)
  })

  
  test('should call the delete function', () => {
    wrapper.find('button').simulate('click')
    expect(handleDelete).toHaveBeenCalledWith(demoTodos[0].id)
  })
  
  
  test('should call the toggle function', () => {
    wrapper.find('label').simulate('click')
    expect(handleToogle).toHaveBeenCalledWith(demoTodos[0].id)
  })
 
  
  test('should show text correctly', () => {
    const text = wrapper.find('label').text()
    expect(text).toBe(`${demoTodos[0].id + 1} - ${demoTodos[0].desc}`)
  })
  

  test('should have the complete class', () => {
    expect(wrapper.find('label').hasClass('complete')).toBeTruthy()
  })
  
})
