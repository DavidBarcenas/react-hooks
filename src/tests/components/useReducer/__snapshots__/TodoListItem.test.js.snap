// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TodoListItem /> testing should render correctly 1`] = `
<li
  className="list-group-item d-flex justify-content-between align-items-center"
>
  <label
    className="complete mb-0"
    onClick={[Function]}
  >
    2
     - 
    Learning React
  </label>
  <button
    className="btn btn-danger btn-sm"
    onClick={[Function]}
  >
    X
  </button>
</li>
`;
